i18n-helpers/src/xgettext.rs

```diff
@@ -41,7 +41,13 @@ fn strip_link(text: &str) -> String {
 
 fn add_message(catalog: &mut Catalog, msgid: &str, source: &str) {
     let sources = match catalog.find_message(None, msgid, None) {
-        Some(msg) => wrap_sources(&format!("{}\n{}", msg.source(), source)),
+        Some(msg) => {
+            if msg.source().contains(source) {
```

Comment:
I think checking for existence should be done per line rather than regular string contains. An example case where the proposed logic would fail will be if source contains entry `some-dir/README.md` and the new source is `README.md`.

Reply:
Good catch @kdarkhan! I think this could be handled nicely by

1. Separating the sources with `\n` when `add_message` is called.
2. Use `lines()` to split the source field back to lines and then deduplicate these lines.
3. Do a call to `wrap_sources` to nicely wrap things.

Both 2 and 3 could be done _after_ adding all messages to the catalog — I think that would give us linear complexity here. Right now, we have some potential for O(n²) complexity here.

Putting a filename per line remove ambiguity about filenames that contain spaces. I'm not actually sure how such filenames ought to be stored in the sources field — feel free to experiment a bit with `xgettext` if you are curious!

Reply:
Thanks for the outline @mgeisler! I pushed an update along those lines. What I have assumes that duplicate sources are consecutive. Are there cases where that doesn't hold?

Reply:
> What I have assumes that duplicate sources are consecutive. Are there cases where that doesn't hold?

I think it should hold right now — we walk over the contents of the book in the order given by [`Book::iter`](https://docs.rs/mdbook/latest/mdbook/book/struct.Book.html#method.iter). It says that it's doing a depth-first traversal on the files of the book, so it ought to do the right thing!

Reply:
This change looks good to me.

Here is the diff of the generated `messages.pot` from `comprehensive-rust` repo.

https://gist.github.com/kdarkhan/9e270ae1e5842abd10f826dcaa5940aa